### ISSUE TITLE
Questions regarding the new YUV 4:4:4 mode triggering/enabling VRR

### ISSUE THREAD

**Issue by charnet3d** [moonlight-stream/moonlight-qt#1424]
Hi,
So after having upgraded to version 6.1 I noticed that there was this shiny new feature : YUV 4:4:4 mode which enhances the color accuracy but requiring much more bandwidth than normal.

I updated to the beta version of Sunshine and went on to test this on my samsung 4k 144hz Oled which supports VRR. After a bit of tinkering and confirming that my LAN can stream 200+ mbps reliably I noticed on the TV OSD that the refresh rate is actually following the FPS in game (with quite a bit of fluctuations), which made me go nuts thinking that something was finally added that enables VRR support, however I found nothing on the feature lists and commit/PR history.
Then I checked if it is actually this new yuv444 mode that did the trick and it was. As when I disable it the TV stays at a constant 144hz.

Are you guys aware of this behavior, what's causing it underneat? is it maybe a new rendering method that was required by yuv444 that did that or what ? Could this be considered a good enough VRR implementation?

I tried to do some A/B testing and see if this decreases frame skipping or is it maybe just placebo in my eyes, while the stutter wasn't visibly gone, I did notice that there's a lot less tearing when I have VSync OFF in both cases. The test I did was using [this utility](https://github.com/Nixola/VRRTest), I set the FPS to keep going up and down between 30 and 144.

Here are the videos I captured with my phone, you can see the refresh rate in the gamebar below and the current FPS in the top left:

[Fixed Refresh - AV1 without yuv444](https://www.youtube.com/watch?v=6qxX0znvFao)
[Variable Refresh - AV1 with yuv444](https://www.youtube.com/watch?v=j66C_YMwSac)

Can you guys do a test to see if this is something to actually celebrate or I'm just in a waking dream about VRR being a thing :D

**Comment by cgutman** [comment]
Streaming YUV 4:4:4 will use the Vulkan-based renderer, which is new (for Windows) in v6.1.0 and could definitely behave differently than the D3D11 renderer. The one exception is that systems with Intel GPUs will use the existing D3D11 renderer (since Intel lacks working support for Vulkan video decoding). I wonder if the reason that your behaviors differs on the 4090 client is that it has an Intel CPU, so Moonlight is using D3D11 on the Intel iGPU for decoding instead of Vulkan on the 4090.

D3D11 is supposed to be able to also have functional VRR support with the V-Sync toggle off, but I'm not sure why that's not working in your case. We set the tearing flags that [Microsoft recommends for VRR](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/variable-refresh-rate-displays) when V-Sync is off.

Anyway, if you want to use the Vulkan renderer all the time, you can set the `D3D11VA_ENABLED=0` environment variable to turn off D3D11 decoding. That should even work on the 4090 system if it is indeed prefering the iGPU over the 4090.

**Comment by charnet3d** [comment]
No I didn't use an Intel iGPU. The 4090 rig has a 5950X and the other has a 3900X. So it's weird that they were different.

I will try that env variable once I'm home maybe something was off for some reason. I'll also double check my sunshine servers for any differences in configuration.

**Comment by charnet3d** [comment]
Okey I figured it out.

First thing I did was a DDU and a driver update to the newest version.
First remark was that I could not notice any tearing whatsoever on this 1440p170hz monitor with whatever settings I chose in moonlight.
So I decided to check with my 2nd monitor, a 1080p144hz, usually hooked up with hdmi as that cable is longer but not supporting vrr. I used the DP cable from main monitor, enabled vrr and checked again in moonlight, then everything worked as expected, similar to the TV: Tearing without yuv444, vrr working with yuv444 and the Hz changing in the OSD and the env variable also worked as expected, with a restart of moonlight with every toggle.
Now I switch back to the 170hz, this time I got the idea to try using the NATIVE res, similar to the test I just did on the FHD mon. I usually always use 2.25x DLDSR and 125% DPI, giving crisper detail and slightly more working space. Anyway with the native res it finally gave me tearing (the only time you'd be happy to see it 😄), and when vulkun renderer is on, either with yuv444 or the env variable the VRR is showing correctly on the OSD.

So in conclusion DSR somehow forces VRR but in a weird state where it could be working (as no tearing is visible) and not reporting back to the monitor ...
Thanks @cgutman for your feedback, at least I know about that variable in case I want the vrr but without a crazy bandwidth increase.

Still would love to know what other people could add to this subject. I'll keep this issue open for a little while longer.

**Comment by rramnel** [comment]
I am using a RTX 4080 as the host and I noticed that my client (running AMD 780m iGPU) is only able to do software decoding with YUV 444 enabled. On my older MacBook M1 Pro I am able to use YUV444 with hardware decoding. 

Is there any setting or option I would need to enable to get hardware decoding to work with my AMD laptop with YUV444 enabled?

**Comment by cgutman** [comment]
> Is there any setting or option I would need to enable to get hardware decoding to work with my AMD laptop with YUV444 enabled?

Unfortunately, no current AMD GPUs support YUV 4:4:4 encoding or decoding. It's a hardware limitation.

**Comment by acrontum-revah** [comment]
> Okey I figured it out.
> 
> First thing I did was a DDU and a driver update to the newest version. First remark was that I could not notice any tearing whatsoever on this 1440p170hz monitor with whatever settings I chose in moonlight. So I decided to check with my 2nd monitor, a 1080p144hz, usually hooked up with hdmi as that cable is longer but not supporting vrr. I used the DP cable from main monitor, enabled vrr and checked again in moonlight, then everything worked as expected, similar to the TV: Tearing without yuv444, vrr working with yuv444 and the Hz changing in the OSD and the env variable also worked as expected, with a restart of moonlight with every toggle. Now I switch back to the 170hz, this time I got the idea to try using the NATIVE res, similar to the test I just did on the FHD mon. I usually always use 2.25x DLDSR and 125% DPI, giving crisper detail and slightly more working space. Anyway with the native res it finally gave me tearing (the only time you'd be happy to see it 😄), and when vulkun renderer is on, either with yuv444 or the env variable the VRR is showing correctly on the OSD.
> 
> So in conclusion DSR somehow forces VRR but in a weird state where it could be working (as no tearing is visible) and not reporting back to the monitor ... Thanks [@cgutman](https://github.com/cgutman) for your feedback, at least I know about that variable in case I want the vrr but without a crazy bandwidth increase.
> 
> Still would love to know what other people could add to this subject. I'll keep this issue open for a little while longer.

I'm having a similar issue to the one you were having in the VRR video you posted, where my TV's refresh rate goes wild whenever i'm using Moonlight, see here I described the issue as best as I could:
https://github.com/moonlight-stream/moonlight-qt/issues/1545

When you say you figured it out, did you also manage to figure out this issue where the TV's refresh rate doesn't align with Moonlight's reported FPS?
It would make sense to me that if Moonlight is reporting 75 FPS, then my VRR TV's refresh rate should also hover around 75 Hz.

**Comment by elisoftli** [comment]
> Okey I figured it out.
> 
> First thing I did was a DDU and a driver update to the newest version. First remark was that I could not notice any tearing whatsoever on this 1440p170hz monitor with whatever settings I chose in moonlight. So I decided to check with my 2nd monitor, a 1080p144hz, usually hooked up with hdmi as that cable is longer but not supporting vrr. I used the DP cable from main monitor, enabled vrr and checked again in moonlight, then everything worked as expected, similar to the TV: Tearing without yuv444, vrr working with yuv444 and the Hz changing in the OSD and the env variable also worked as expected, with a restart of moonlight with every toggle. Now I switch back to the 170hz, this time I got the idea to try using the NATIVE res, similar to the test I just did on the FHD mon. I usually always use 2.25x DLDSR and 125% DPI, giving crisper detail and slightly more working space. Anyway with the native res it finally gave me tearing (the only time you'd be happy to see it 😄), and when vulkun renderer is on, either with yuv444 or the env variable the VRR is showing correctly on the OSD.
> 
> So in conclusion DSR somehow forces VRR but in a weird state where it could be working (as no tearing is visible) and not reporting back to the monitor ... Thanks [@cgutman](https://github.com/cgutman) for your feedback, at least I know about that variable in case I want the vrr but without a crazy bandwidth increase.
> 
> Still would love to know what other people could add to this subject. I'll keep this issue open for a little while longer.

I'm having a similar issue to the one you were having in the VRR video you posted, where my TV's refresh rate goes wild whenever i'm using Moonlight, see here I described the issue as best as I could:
https://github.com/moonlight-stream/moonlight-qt/issues/1545

When you say you figured it out, did you also manage to figure out this issue where the TV's refresh rate doesn't align with Moonlight's reported FPS?
It would make sense to me that if Moonlight is reporting 75 FPS, then my VRR TV's refresh rate should also hover around 75 Hz.

**Comment by africsnail** [comment]
For me VRR seems to only work using the YUV mode and (weirdly enough) only with V-Sync toggle enabled. When either isn't met, VRR doesn't work and TV's OSD shows 120.

I'm using LG C1 and RTX 3070 Windows 11 Laptop as the client. The host is Windows 11 PC with RTX 4080.

**Comment by africsnail** [comment]
For me VRR seems to work only with the vulkan decoding but it's also way slower using that (50ms+) on my RTX 3070 laptop client. VRR doesn't seem to work with the default decoding (either with vsync off or on)

There is significant tearing with the default decoding and with vsync off. So only usable option for now seems to be the DX decoding with vsync.

Would be glad if VRR receives some proper support as it would be flawless then.

**Comment by ns6089** [comment]
> For me VRR seems to work only with the vulkan decoding

You may need to select "borderless windowed" display mode instead of "fullscreen (recommended)" in moonlight-qt settings, and turn off vsync. Direct3D requires windowed presentation for VRR, but I'm not certain how SDL2 treats `SDL_WINDOW_FULLSCREEN_DESKTOP` and `SDL_WINDOW_FULLSCREEN` flags internally.

**Comment by moi952** [comment]
> Streaming YUV 4:4:4 will use the Vulkan-based renderer, which is new (for Windows) in v6.1.0 and could definitely behave differently than the D3D11 renderer. The one exception is that systems with Intel GPUs will use the existing D3D11 renderer (since Intel lacks working support for Vulkan video decoding). I wonder if the reason that your behaviors differs on the 4090 client is that it has an Intel CPU, so Moonlight is using D3D11 on the Intel iGPU for decoding instead of Vulkan on the 4090.
> 
> D3D11 is supposed to be able to also have functional VRR support with the V-Sync toggle off, but I'm not sure why that's not working in your case. We set the tearing flags that [Microsoft recommends for VRR](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/variable-refresh-rate-displays) when V-Sync is off.
> 
> Anyway, if you want to use the Vulkan renderer all the time, you can set the `D3D11VA_ENABLED=0` environment variable to turn off D3D11 decoding. That should even work on the 4090 system if it is indeed prefering the iGPU over the 4090.

Hello,

Given that the smooth / VRR-like behavior only happens when YUV 4:4:4 is enabled (which forces the Vulkan renderer), it strongly suggests that Vulkan presentation timing is the key factor allowing this behavior.

Have any tests been done to confirm whether this VRR-like behavior is specifically tied to the Vulkan renderer rather than the color format itself?

If that’s the case, would it be feasible to expose an option (for example “VRR” or “Force Vulkan renderer”) that explicitly forces Vulkan usage, instead of indirectly relying on YUV 4:4:4 to trigger it?

This would help users benefit from the smoother presentation without being constrained by bandwidth or color format requirements.

Thanks for the great work.

**Comment by ns6089** [comment]
Vulkan and D3D11VA backends in moonlight-qt share the same frame-pacing logic. You can force vulkan backend already through environment variable, and I don't think a dedicated option for VRR could fly. Mostly because it's not controlled on application level, but on system and driver level.

If cgutmat doesn't get to this first, somewhere in the future I will try to dig into the state of VRR in D3D11VA regarding whether or not it could be improved.

For now, if you're on laptop and having problems with stream smoothness, I recommend making sure this option is disabled in Windows display settings:
<img width="718" height="246" alt="Image" src="https://github.com/user-attachments/assets/375637f8-8872-4d54-b192-daa5969fa983" />
It doesn't do what you would expect from freesync and changes refresh rate in large steps, something like 30 or 60 Hz.

**Comment by ns6089** [comment]
@cgutman Just made a few tests (against my better judgement). I'm pretty sure we're not supposed to use DXGI_PRESENT_ALLOW_TEARING together with SyncInterval 0. I get tearing in this case (both fullscreen and borderless), but with SyncInterval 1 tearing disappears and VRR starts to function as expected. UFO Test streamed from 50Hz display to 60Hz display looks stutter-free at least.

**Comment by ns6089** [comment]
But again VRR works just fine for me through Windows-level compatibility layer. Same 50Hz to 60Hz test is smooth in Fullscreen + VSync + Frame pacing. So I'm not sure whether it's worth the time to complicate swapchain logic (and whether it makes sense to keep the option that disables VSync).

**Comment by ns6089** [comment]
> I'm pretty sure we're not supposed to use DXGI_PRESENT_ALLOW_TEARING together with SyncInterval 0

Actually, disregard this. DXGI_PRESENT_ALLOW_TEARING with SyncInterval 1 gives error on (some?) other devices. And VRR on laptop screens is weird in general:
<img width="523" height="117" alt="Image" src="https://github.com/user-attachments/assets/cdba569f-9d2c-492d-8895-ebd48515ed71" />
<img width="454" height="120" alt="Image" src="https://github.com/user-attachments/assets/89497043-52a2-4c67-bec6-f3b4e94f06f5" />
(what do you mean by "Suppported by driver"!?)

**Comment by krblaz** [comment]
I noticed the same behaviour on my end with both the latest  nightly (d865c77) and 6.1.0 on a PC with Win 11 25H2 and 3060 TI. When using D3D11VA with borderless window with vsync toggle to off VRR doesn’t engage when streaming but it works correctly in the UI. When using Vulkan renderer via environment variable VRR works correctly but rendering latency is about 8 ms as opposed to sub 1 ms.

**Comment by cgutman** [comment]
VRR is working fine for me here on my Arc A770 on Win11 25H2 on both d865c77975d1548a095e3f901ed4f82911db6342 and 6.1.0. It doesn't seem to matter whether I use Vulkan or D3D11, windowed/borderless/fullscreen, or V-Sync enabled/disabled. Every combination works (confirmed using my monitor's frame rate overlay).

The D3D11 renderer does everything by the book [according to Microsoft's VRR documentation](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/variable-refresh-rate-displays). We set `DXGI_SWAP_CHAIN_FLAG_ALLOW_TEARING` and `DXGI_PRESENT_ALLOW_TEARING` and use SyncInterval 0, just as Microsoft says. Plus if we didn't, V-sync off wouldn't work in general (steps to enable VRR are exactly the same as how to disable V-Sync).

This could just be an Nvidia driver bug. Is anyone who is having VRR problems _not_ running an Nvidia GPU? If you're on Nvidia, do you have G-Sync set to "Full screen mode" or "Windowed and full screen mode"?

**Comment by krblaz** [comment]
I've tried it again on a different PC and display altogether which has both Nvidia GPU and Ryzen iGPU. When the display is connected to the Nvidia GPU VRR doesn't work in any mode. I made sure VRR is enabled in control pannel for both fullscreen and borderless mode. When switching the display to the AMD iGPU VRR starts working correctly. So the issue seem to lie with Nvidia when using D3D11VA .

**Comment by ns6089** [comment]
My tearing was on Intel, but also laptop Intel and since they've overcomplicated this area recently with [display-side presentation queue](https://learn.microsoft.com/en-us/windows-hardware/drivers/display/hardware-flip-queue) it's hard to tell who's to blame there. Also everything works fine with vsync on, so let's pretend I'm an outlier.

@krblaz Just for confirmation, you didn't apply any of the "Disable MPO" registry tweaks that were relatively popular? They may interfere with how latest Windows version handles independent flip.

**Comment by krblaz** [comment]
No I haven't changed anything in regards to how Windows handles rendering. Besides the behaviour is the same on two different PCs with two different Windows installs.

**Comment by cgutman** [comment]
I tracked this down to Nvidia driver behavior. I believe Nvidia is trying to avoid enabling VRR for video content, and the fact that we share our `ID3D11Device` between FFmpeg and the D3D11 renderer means that Nvidia's driver "sees" that we're decoding video and triggers internal logic to force VRR off. Notably, this happens both without `DXGI_PRESENT_ALLOW_TEARING` (which is an acceptable design decision by Nvidia, as the spec doesn't promise VRR without that flag) and with `DXGI_PRESENT_ALLOW_TEARING` (which is a bug, as it directly contradicts the [Microsoft's VRR enablement procedure](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/variable-refresh-rate-displays) and even [Nvidia's own documentation](https://developer.nvidia.com/blog/advanced-api-performance-swap-chains/)). If I quickly hack the D3D11VA renderer code to let FFmpeg create its own `ID3D11Device`, VRR enables just fine in all scenarios (just like AMD and Intel).

Fortunately, I think this should be easy to fix by using separate D3D11 devices for decoding and rendering and sharing the texture array between the devices. This should also improve performance by allowing decoding and rendering to be done in parallel, though I'll need to check how it performs on non-Nvidia cards (I remember when we switched to using multiple D3D11 devices years ago, it actually hurt AMD GPU performance slightly).

**Comment by cgutman** [comment]
The latest nightly build has separate render and decode devices implemented and VRR is now working on my RTX 3050 test system.

https://nightly.link/moonlight-stream/moonlight-qt/workflows/build/master if you want to test it out.

**Comment by moi952** [comment]
> The latest nightly build has separate render and decode devices implemented and VRR is now working on my RTX 3050 test system.
> 
> https://nightly.link/moonlight-stream/moonlight-qt/workflows/build/master if you want to test it out.

So VRR will work regardless of the options, or does it require enabling something (or a specific VRR option)? 

Will it work on an AMD GPU (RDNA3)? 

Thanks

**Comment by cgutman** [comment]
> So VRR will work regardless of the options, or does it require enabling something (or a specific VRR option)?

It looks like it works regardless of what options are enabled (same as it did for me on Intel). Though technically, the driver only has to enable it when you have v-sync off.

> Will it work on an AMD GPU (RDNA3)?

VRR on AMD and Intel already worked prior to this change, so they shouldn't be impacted at all.

**Comment by moi952** [comment]
> > So VRR will work regardless of the options, or does it require enabling something (or a specific VRR option)?
> 
> It looks like it works regardless of what options are enabled (same as it did for me on Intel). Though technically, the driver only has to enable it when you have v-sync off.
> 
> > Will it work on an AMD GPU (RDNA3)?
> 
> VRR on AMD and Intel already worked prior to this change, so they shouldn't be impacted at all.

With version 6.1.0, it seems that 444 needs to be enabled to switch to Vulkan so that VRR works, right?

**Comment by cgutman** [comment]
As I alluded to in my previous comment, it depends on the GPU vendor. Intel and AMD GPUs enabled VRR on v6.1.0 without requiring Vulkan. It was only Nvidia cards where the Vulkan/4:4:4 distinction mattered in v6.1.0.

In currently nightly builds, Nvidia GPUs behave the same as Intel/AMD. VRR works regardless of your 4:4:4, display mode, or v-sync selection.

**Comment by nebuchadnezza343** [comment]
Just to add another datapoint

Moonlight 6.1 MacOS vs Nightly Moonlight-macOS-8d5072

6.1 triggers VRR although not perfect
Nightly shows 118.6X Hz all the time on my LG C3

HDR wise the 6.1 maxes out at around 400nits and therefore crushes the whites unless I configure the games and windows to that maximum.
Nightly 8d5072 on the other hand maxes out at around 850 nits which is close enough to the 800nits the TV supports (and my main TV connected to the PC - same but smaller) and therefore does not need any recalibration for HDR streaming.

So the best of both worlds would be HDR from Nightly with VRR triggering from 6.1.

All these where done using Cyberpunk and Virtual Desktop second screen to capture frame generation frames properly. (FG is not necessary to see the VRR / HDR difference though)

Best regards!

**Comment by andygrundman** [comment]
> HDR wise the 6.1 maxes out at around 400nits and therefore crushes the whites unless I configure the games and windows to that maximum. Nightly 8d5072 on the other hand maxes out at around 850 nits which is close enough to the 800nits the TV supports (and my main TV connected to the PC - same but smaller) and therefore does not need any recalibration for HDR streaming.

This thread is way off-topic so I might as well repost my instructions for calibrating Mac HDR on an LG TV.

(From a pinned comment in #qt-help on Discord.)
This is one way I have found to calibrate HDR for a Moonlight client on macOS Sequoia with an external display. If you are using an Apple display such as a MacBook Pro, all you need to do is switch to the HDR Video preset and Windows HDR Calibration should be a perfect 1000 nits.

I went and re-checked what I did on my Mini with my C2 for HDR. What I would recommend is to run Moonlight in a window and have the display settings nearby on the side. 

Go into HDR calibration in Windows to the screen where you match the white square. Match it and then play with the brightness slider to see what happens. At 100% Mac brightness, it would match at 420 for me, if I lowered Mac to 50% it went up to the 1900s. I know the real max of my TV is 800 so I moved the Windows slider to 800 and then moved the Mac slider until it was even, about 80% on my system. This is bascially like the XDR mode on a Macbook and as long as you calibrate Windows correctly, it will look correct. The question is how much dynamic range do you want? Mac at 100% and Windows at 420 is still running the display at its max of 800 but you have less range. And if you run the TV darker you can have more dynamic range. I think if you do the 50% brightness it will feel too dark overall, but as long as you know how these sliders work you can play with it to find a good setting. 

Also remember you need to enable this profile in Windows whenever you stream, and then set it back if you use the system directly. For anyone interested in the technical details this video is a good start: https://developer.apple.com/videos/play/wwdc2021/10161/ 

![Image](https://github.com/user-attachments/assets/22d54229-6b28-409c-9334-932bce49b1b0)
